File: 05-image-fades/main.py
#!/usr/bin/env python3
''' Experiment 1 - Scrolling Texture

By Chris Herborth (https://github.com/Taffer)
MIT license, see LICENSE.md for details.
'''

import pygame
import time
import sys

SCREEN_TITLE = 'Experiment 5 - Image Fades'

SCREEN_WIDTH = 1280  # 720p screen
SCREEN_HEIGHT = 720

BLACK = pygame.Color('black')


class Demo:
    def __init__(self, screen):
        self.screen = screen

        self.robot1 = pygame.image.load('resources/character_robot_jump.png').convert_alpha()
        self.rect1 = self.robot1.get_rect()
        self.rect1.top = 100
        self.rect1.left = 100

        self.rect2 = self.robot1.get_rect()
        self.alpha2 = pygame.Surface((self.rect2.width, self.rect2.height))
        self.alpha2.fill(BLACK)
        self.alpha2.set_alpha(255)
        self.rect2.top = 100
        self.rect2.left = 250

        self.alpha = 255

        self.ticks = 0

    def draw(self):
        self.screen.fill(BLACK)

        self.screen.fill(BLACK)
        self.screen.blit(self.robot1, self.rect1)  # Normal sprite

        self.screen.blit(self.robot1, self.rect2)  # Linear fade
        self.screen.blit(self.alpha2, self.rect2)

    def update(self, dt):
        self.ticks += dt

        if self.ticks > 1/60:
            self.ticks -= 1/60
            self.alpha -= 256/120
            if self.alpha < 0:
                self.alpha = 255
            self.alpha2.set_alpha(self.alpha)


def main():
    pygame.init()

    screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
    pygame.display.set_caption(SCREEN_TITLE)

    demo = Demo(screen)

    now = time.time()
    dt = 0

    while True:
        demo.draw()
        pygame.display.flip()

        dt = time.time() - now
        now = time.time()

        demo.update(dt)

        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                sys.exit()
            elif event.type == pygame.KEYDOWN and event.key == pygame.K_ESCAPE:
                sys.exit()


if __name__ == '__main__':
    main()

File: 05-image-fades/main.lua
-- Experiment 5 - Image Fades
--
-- By Chris Herborth (https://github.com/Taffer)
-- MIT license, see LICENSE.md for details.

-- All the stuff we've loaded already.
gameResources = {
    fonts = {},
    images = {}
}

-- Current state of the game.
gameState = {
    sprite_x = 100, -- Draw the normal sprite here.
    sprite_y = 100,

    sprite_dx = 150, -- Delta x for subsequent sprites.

    alpha = 0, -- Alpha channel for fades. 0 to 1 in 1 second.

    wait = 0,     -- wait 1 second, then start alpha over at 0
    wait_time = 1
}

-- Love callbacks.
function love.load()
    math.randomseed(os.time())
    love.graphics.setDefaultFilter('nearest', 'nearest')

    local gameResources = gameResources
    gameResources.images.robot = love.graphics.newImage('resources/character_robot_jump.png')
end

function love.draw()
    local gameResources = gameResources
    local gameState = gameState

    -- Background
    love.graphics.clear(0, 0, 0, 1)

    -- Normal sprite
    love.graphics.setColor(1, 1, 1, 1) -- White = "Draw images normally."
    love.graphics.draw(gameResources.images.robot, gameState.sprite_x, gameState.sprite_y)

    -- Linear fade.
    local sprite2_x = gameState.sprite_x + gameState.sprite_dx
    local sprite2_y = gameState.sprite_y
    love.graphics.setColor(1, 1, 1, gameState.alpha)
    love.graphics.draw(gameResources.images.robot, sprite2_x, sprite2_y)

    -- sin() fade
    sprite2_x = sprite2_x + gameState.sprite_dx
    local rads = gameState.alpha * (90 * math.pi / 180)
    love.graphics.setColor(1, 1, 1, math.sin(rads))
    love.graphics.draw(gameResources.images.robot, sprite2_x, sprite2_y)
end

function love.update(dt)
    local gameState = gameState

    -- Alpha += dt/2 means fade-in takes 2 seconds.
    gameState.alpha = gameState.alpha + dt / 2

    if gameState.alpha > 1 then
        gameState.alpha = 1

        gameState.wait = gameState.wait + dt
        if gameState.wait > gameState.wait_time then
            gameState.alpha = 0
            gameState.wait = 0
        end
    end
end

-- Event generation.
function love.keypressed(key)
    if key == 'escape' then
        love.event.quit()
    end
end
